File: 09.Greed贪心/leetcode_1727_largestSubmatrix.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int largestSubmatrix(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<int> rod(n, 0);
        int res = 0;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                rod[j] = matrix[i][j]==0 ? 0 : rod[j]+1;
            }

            vector<int> sortedRod = rod;
            sort(sortedRod.begin(), sortedRod.end());
            for (int k=0; k<n; k++){
                res = max(res, (n-k)*sortedRod[k]);
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<int>> matrix = {{0,0,1},{1,1,1},{1,0,1}};
    cout << s.largestSubmatrix(matrix) << endl;
}

File: 09.Greed贪心/leetcode_1727_largestSubmatrix.py
from typing import List

class Solution:
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        # 在循环中，rods[j]始终维护第j列从当前行往上数连续1的个数，那么只需要对这些1的个数排序，即可贪心地求得以当前行为底边的最大矩形面积
        m, n = len(matrix), len(matrix[0])
        res = 0
        rods = [0]*n
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    rods[j] = 0
                else:
                    rods[j] += 1
            
            sortedRods = sorted(rods)
            for k in range(n):
                res = max(res, (n-k)*sortedRods[k])
        return res

if __name__ == "__main__":
    s = Solution()
    matrix = [[0,0,1],[1,1,1],[1,0,1]]
    print(s.largestSubmatrix(matrix))
